nixos/modules/security/gnupg.nix

```diff
@@ -0,0 +1,270 @@
+          description = ''
+            Shell script taking the deciphered secret on its standard input
+            and which must put on its standard output
+            the actual secret material to be installed.
+            This allows to decorate the secret with non-secret bits.
+          '';
+        };
+        path = lib.mkOption {
+          type = types.str;
+          default = name;
+          apply = p: if match "^/.*" p == null then "/run/keys/gnupg/"+p+"/file" else p;
+          description = ''
+            The path on the target system where the secret is installed to.
+            Default to the name of the secret,
+            prefixed by <filename>/run/keys/gnupg/</filename>
+            and suffixed by <filename>/file</filename>,
+            if non-absolute.
+          '';
+        };
+        service = lib.mkOption {
+          type = types.str;
+          default = "secret-" + escapeUnitName name + ".service";
+          description = ''
+            The name of the systemd service.
+            Useful to put constraints like <literal>after</literal> or <literal>wants</literal>
+            into services requiring this secret.
+          '';
+        };
+        postStart = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = "systemctl reload nginx.service";
+          description = ''
+            Commands to run after new secrets go live.
+            Typically the web server and other servers using secrets
+            need to be reloaded.
+          '';
+        };
+        postStop = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = ''shred -u "$secret_file"'';
+          description = ''
+            Commands to run after stopping the service.
+            Typically removing a persistent secret.
+            For convenience the <literal>path</literal> of the secret
+            is provided in the shell variable <literal>secret_file</literal>.
+          '';
+        };
+      };
+    }));
+  };
+  enableGpgAgent = lib.mkEnableOption ''gpg-agent for decrypting secrets.
+
+    Otherwise, you'll have to forward an <literal>agent-extra-socket</literal>:
+    <screen>
+    <prompt>$ </prompt>ssh -nNT root@example.org -o StreamLocalBindUnlink=yes -R /var/lib/gnupg/S.gpg-agent:$(gpgconf --list-dirs agent-extra-socket)
+    </screen>
+  '' // {default = true; example = false;};
+  gpgAgentFlags = lib.mkOption {
+    type = types.listOf types.str;
+    default = [
+      "--default-cache-ttl" "600"
+      "--max-cache-ttl" "7200"
+    ];
+    description = ''
+      Extra flags passed to the <literal>gpg-agent</literal>
+      used to decrypt secrets.
+    '';
+  };
+};
+config = lib.mkIf (cfg.secrets != {}) {
+  # Because /run/user/0 is wiped out by pam_systemd when root logouts,
+  # systemd.services.gpg-agent cannot put its socket in
+  # the path expected by gpg: /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
```

Comment:
Maybe I'm missing something obvious but why the `d.6qoenf9br6fajbkknuz1i6ts` name?

Reply:
It's the standard socket path used by GnuPG, derived from the given `--homedir`. With the workarounds used here, `gpg`/`gpg-agent` cannot create this standard socket and default to `<homedir>/S.gpg-agent`. This is wanted here to keep the `gpg-agent` alive, because when `root` logouts its `/run/user/0/` is removed and `gpg-agent` would no longer be accessible, detect that and kill itself.
See also https://github.com/gpg/gnupg/commit/aab8a0b05292b0d06e3001a0b289224cb7156dbd